src/librustc_codegen_llvm/back/link.rs

```diff
@@ -123,33 +121,33 @@ pub(crate) fn link_binary(sess: &Session,
                           codegen_results: &CodegenResults,
                           outputs: &OutputFilenames,
                           crate_name: &str) -> Vec<PathBuf> {
-    let mut out_filenames = Vec::new();
-    for &crate_type in sess.crate_types.borrow().iter() {
+    let out_filenames = sess.crate_types.borrow().iter().filter_map(|&crate_type| {
         // Ignore executable crates if we have -Z no-codegen, as they will error.
         let output_metadata = sess.opts.output_types.contains_key(&OutputType::Metadata);
-        if (sess.opts.debugging_opts.no_codegen || !sess.opts.output_types.should_codegen()) &&
-           !output_metadata &&
-           crate_type == config::CrateType::Executable {
-            continue;
+        if (sess.opts.debugging_opts.no_codegen || !sess.opts.output_types.should_codegen())
+           && !output_metadata
+           && crate_type == config::CrateType::Executable
+        {
+            return None
         }
 
         if invalid_output_for_target(sess, crate_type) {
            bug!("invalid output type `{:?}` for target os `{}`",
                 crate_type, sess.opts.target_triple);
         }
-        let mut out_files = link_binary_output(sess,
-                                               codegen_results,
-                                               crate_type,
-                                               outputs,
-                                               crate_name);
-        out_filenames.append(&mut out_files);
-    }
+
+        let out_files = link_binary_output(sess,
+                                           codegen_results,
+                                           crate_type,
+                                           outputs,
+                                           crate_name);
+        Some(out_files)
+    })
+    .flatten().collect::<Vec<_>>();
```

Comment:
This seems less clear. If we're going to make this change, I'd prefer that we instead just change the `append` to `extend` and leave it otherwise the same.

Reply:
Agreed.